src/python/arcor2_arserver/lock.py

```diff
@@ -0,0 +1,259 @@
+
+            if unlocked and lock_record.is_empty():
+                del self._locked_objects[root_id]
+
+        return unlocked
+
+    async def write_lock(self, obj_id: str, owner: str, lock_tree: bool = False) -> bool:
+
+        root_id = self.get_root_id(obj_id)
+
+        async with self._lock:
+            lock_record = self._get_lock_record(root_id)
+            locked = lock_record.write_lock(obj_id, owner, lock_tree)
+
+            if not locked and lock_record.is_empty():
+                del self._locked_objects[root_id]
+
+        return locked
+
+    async def write_unlock(self, obj_id: str, owner: str) -> bool:
+
+        root_id = self.get_root_id(obj_id)
+
+        async with self._lock:
+            lock_record = self._get_lock_record(root_id)
+            unlocked = lock_record.write_unlock(obj_id, owner)
+
+            if unlocked and lock_record.is_empty():
+                del self._locked_objects[root_id]
+
+        return unlocked
+
+    async def is_write_locked(self, obj_id: str, owner: str) -> bool:
+
+        root_id = self.get_root_id(obj_id)
+
+        async with self._lock:
+            lock_record = self._locked_objects.get(root_id)
+            if not lock_record:
+                return False
+
+            return obj_id in lock_record.write and owner in lock_record.write[obj_id].owners
+
+    async def is_read_locked(self, obj_id: str, owner: str) -> bool:
+
+        root_id = self.get_root_id(obj_id)
+
+        async with self._lock:
+            lock_record = self._locked_objects.get(root_id)
+            if not lock_record:
+                return False
+
+            return obj_id in lock_record.read and owner in lock_record.read[obj_id].owners
+
+    async def get_locked_roots_count(self) -> int:
+
+        async with self._lock:
+            return len(self._locked_objects)
+
+    def get_root_id(self, obj_id):
+        """Retrieve root object id for given object"""
+
+        if self.scene:
+            # TODO support scene object hierarchy
+            return obj_id
+
+        elif self.project:
+            obj = self.project.action_point(obj_id)
+            return obj.parent if obj.parent else obj_id
+
+        elif obj_id in [self.SCENE_NAME, self.PROJECT_NAME]:
+            return obj_id
+
+        else:
+            raise KeyError(f"Unknown object '{obj_id}'")
+
+    def _get_lock_record(self, root_id: str) -> "Lock.LockedObject":
+        """Create and/or retrieve lock record for root_id"""
+
+        assert self._lock.locked()
+
+        if root_id not in self._locked_objects:
+            self._locked_objects[root_id] = Lock.LockedObject()
+
+        return self._locked_objects[root_id]
+
+    async def get_lock(self) -> AsyncGenerator[asyncio.Lock, None]:
+        """
+        Get lock for data structure
+        Method should be used for operation with whole scene/project, no others
+        """
+        yield self._lock
+
+    async def clean_locks(self):
```

Comment:
Tady by to asi chtělo vložit nějaký sleep? Teď to asi vytěžuje procesor na 100%, ne?